File: genprime.cpp
#include <cmath>
#include <cstdio>
#include <cstdlib>
#include <vector>
#include <sys/time.h>

typedef unsigned long prime_t;

std::vector<prime_t> primes;

bool isprime(prime_t x)
{
	prime_t lim, y;
	if (x < 2)
		return false;
	if (x == 2)
		return true;
	if (x % 2 == 0)
		return false;
	if (x < 9)
		return true;
	if ((x + 1) % 6 != 0)
		if ((x - 1) % 6 != 0)
			return false;
	lim = (prime_t)(sqrt((double)x) + 1.0f);
	y = 1;
	for (std::vector<prime_t>::iterator iter = primes.begin();
		 iter != primes.end();
		 iter++) {
		y = *iter;
		if (y >= lim)
			return true;
		if (x % *iter == 0)
			return false;
	}
	for (; y < lim; y += 2)
	{
		if (x % y == 0)
			return false;
	}
	return true;
}

prime_t genprime(prime_t max)
{
	prime_t count = 0,
		current = 1;
	while (count < max)
	{
		if (isprime(current)) {
			if (primes.size() < 1 || primes.back() < current)
				primes.push_back(current);
			count++;
		}
		current++;
	}
	return current - 1;
}

int main(int argc, char **argv)
{
	prime_t start = argc > 1 ? atol(argv[1]) : 0,
		stop = argc > 2 ? atol(argv[2]) + 1 : 0,
		x, last;
	struct timeval begin, end;
	double duration;
	for (x = start; x < stop; x += start)
	{
		gettimeofday(&begin, NULL);
		last = genprime(x);
		gettimeofday(&end, NULL);
		duration = (double)(end.tv_sec - begin.tv_sec) +
			((double)(end.tv_usec) - (double)(begin.tv_usec)) / 1000000.0;
		printf ("Found %8lu primes in %10.5f seconds (last was %10lu)\n",
			x, (float)duration, last);
	}
	return 0;
}

File: genprime.rs
extern crate num;
extern crate time;

use std::env;

use num::iter;

fn isprime(x: u64) -> bool {
    let lim: u64;

    if x < 2 {
        return false;
    }
    if x == 2 {
        return true;
    }
    if x % 2 == 0 {
        return false;
    }
    if x < 9 {
        return true;
    }
    if (x + 1) % 6 != 0 {
        if (x - 1) % 6 != 0 {
            return false;
        }
    }
    lim = ((x as f64) + 1.0f64).sqrt() as u64;
    for y in iter::range_step_inclusive(3, lim, 2) {
        if x % y == 0 {
            return false;
        }
    }

    return true;
}

fn genprime(max: u64) -> u64 {
    let mut   count: u64 = 0;
    let mut current: u64 = 1;

    while count < max {
        if isprime(current) {
            count += 1;
        }
        current += 1;
    }

    return current - 1;
}

fn main() {
    let  args: Vec<String> = env::args().collect();
    let start: u64;
    let  stop: u64;
    let mut  last: u64;
    let mut begin: f64;
    let mut   end: f64;

    start = if args.len() > 1 {
        args[1].parse().unwrap()
    } else {
        0
    };

    stop = if args.len() > 2 {
        args[2].parse().unwrap()
    } else {
        0
    };

    for x in iter::range_step_inclusive(start, stop, start) {
        begin = time::precise_time_s();
        last = genprime(x);
        end = time::precise_time_s();
        println!("Found {:8} primes in {:10.5} seconds (last was {:10})", x, end - begin, last);
    }
}
